src/NServiceBus.Core/Hosting/StartupDiagnostics/HostStartupDiagnostics.cs

```diff
@@ -0,0 +1,100 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.IO;
+    using System.Threading.Tasks;
+    using Features;
+    using Logging;
+    using Settings;
+    using SimpleJson;
+
+    class HostStartupDiagnostics : Feature
+    {
+        public HostStartupDiagnostics()
+        {
+            EnableByDefault();
+        }
+
+        protected internal override void Setup(FeatureConfigurationContext context)
+        {
+            var settings = context.Settings;
+
+            var diagnosticsWriter = GetDiagnosticsWriter(settings);
+
+            context.RegisterStartupTask(new WriteStartupDiagnostics(diagnosticsWriter, settings));
+        }
+
+
+        static HostDiagnosticsWriter GetDiagnosticsWriter(ReadOnlySettings settings)
+        {
+            if (settings.TryGet<HostDiagnosticsWriter>(out var diagnosticsWriter))
+            {
+                return diagnosticsWriter;
+            }
+
+            if (!settings.TryGet<string>(HostDiagnosticsConfigurationExtensions.DiagnosticsRootPathKey, out var diagnosticsRootPath))
+            {
+                diagnosticsRootPath = Path.Combine(DefaultFactory.FindDefaultLoggingDirectory(), ".diagnostics");
+            }
+
+            if (!Directory.Exists(diagnosticsRootPath))
+            {
+                Directory.CreateDirectory(diagnosticsRootPath);
+            }
+
+            var startupDiagnoticsFileName = "startup-configuration.txt";
+            var startupDiagnoticsFilePath = Path.Combine(diagnosticsRootPath, startupDiagnoticsFileName);
+
+            if (File.Exists(startupDiagnoticsFilePath))
+            {
+                File.Delete(startupDiagnoticsFilePath);
+            }
+
+            return new HostDiagnosticsWriter(data => AsyncFile.WriteText(startupDiagnoticsFilePath, data));
+        }
+
+        class WriteStartupDiagnostics : FeatureStartupTask
+        {
+            public WriteStartupDiagnostics(HostDiagnosticsWriter diagnosticsWriter, ReadOnlySettings settings)
+            {
+                this.diagnosticsWriter = diagnosticsWriter;
+                this.settings = settings;
+            }
+
+            protected override async Task OnStart(IMessageSession session)
+            {
+                try
+                {
+                    var data = "";
+
+                    foreach (var section in settings.Get<StartupDiagnosticEntries>().Entries)
```

Comment:
Is it not possible to serialize the whole dictionary?

Reply:
I tried and it gave this:

```
{
   [
      {
         "Name":"Routing",
         "Data":{
            "UnicastRoutingTable":{

            },
            "DistributionPolicy":{

            },
            "EndpointInstances":{

            },
            "Publishers":{

            },
            "EnforceBestPractices":true
         }
      },
      {
         "Name":"HostInformation",
         "Data":{
            "HostId":"e3342552-4a13-b662-aa77-4461b333c85b",
            "DisplayName":"ANDREAS2015",
            "Properties":{
               "Machine":"ANDREAS2015",
               "ProcessID":"26440",
               "UserName":"andreas.ohlund",
               "PathToExecutable":"C:\\Users\\andreas.ohlund\\AppData\\Local\\JetBrains\\Installations\\ReSharperPlatformVs15_c3a37fcb_000\\JetBrains.ReSharper.TaskRunner.CLR45.x64.exe"
            }
         }
      }
   ]
}
```

are we happy with that? (`Name` and `Data` can be renamed)

Reply:
call todictionary on the entries. with name as the key and data as the value. then serialize that dictionary